### ISSUE TITLE
How to create my own DeviceLocationProvider?

### ISSUE THREAD

**Issue by derekdominoes** [mapbox/mapbox-unity-sdk#118]
I would like a DeviceLocationProvider that, for instance, reads coordinates from a serial port or an EditorLocationProvider that can be moved around from the keyboard.

The LocationProviderFactory accepts concrete references to both a EditorLocationProvider and DeviceLocationProvider instead of ILocationProvider. These two classes do not seemed designed for subclassing either. Thus, it isn't clear to me how to provide my own versions of these to the factory without a wholesale replacement of their code.  Maybe I'm missing something here...?

**Comment by mihakrajnc** [comment]
I had a similar case where I needed to provide location updates from a GPX file. The way I did it was with a `CustomLocationProviderFactory`, something like this:

```csharp
public class CustomLocationProviderFactory : MonoBehaviour
{
    [SerializeField] private GpxLocationProvider gpxProvider;
    [SerializeField] private GpsLocationProvider gpsProvider;
    [SerializeField] private StaticLocationProvider staticLocationProvider;

    private ILocationProvider defaultProvider;

    public ILocationProvider DefaultProvider
    {
        get
        {
            if (defaultProvider == null)
            {
                switch (GameManager.Instance.LocationType)
                {
                    case LocationType.GPS:
                        defaultProvider = gpsProvider;
                        break;
                    case LocationType.GPX:
                        defaultProvider = gpxProvider;
                        break;
                    case LocationType.STATIC:
                        defaultProvider = staticLocationProvider;
                        break;
                    default:
                        throw new ArgumentOutOfRangeException();
                }
            }

            return defaultProvider;
        }
    }
}
```

I then used this factory in my player position controller and anywhere else I needed it. Of course, you could modify Mapbox's scripts, but I wouldn't recommend it, as any update will overwrite your code.

You'll also need to implement your own `ILocationProvider` (e.g. to read from a serial port, for instance).

IMO Mapbox's `LocationProviderFactory` and the scripts that come with it should be part of the examples, or rewritten to provide better extendability, as they aren't really that useful past the specific cases they were made for.

**Comment by david-rhodes** [comment]
@mihakrajnc you are correct--the `factory` itself is more of an example of how to use the backing `ILocationProvider` interface. In fact, you don't need a factory at all, if you write your own provider that knows how to mock location in the editor.

We'll probably expand on this in the future, but for now--just modify as you need (possibly in a new class so future updates don't break anything for you).
